### DISCUSSION TITLE
[Deployment] Need help with error while deploying contract on layer 2(zksync testnet)

### DISCUSSION THREAD

**Discussion by kientt265** [zkSync-Community-Hub/zksync-developers#683] [Deployments] [answered]
### Environment

Testnet

### L2 block number

n/a

### zkSolc Version

10.8.1

### zksync-ethers Version

10.8.1

### Minimal Reproducible Code or Repo Link

https://github.com/kientt265/zksync-project.git

### Does this work on other EVMs? (If yes, please list at least 1 of them)

no

### Description of What Your Contract Does

I want to build an interactive web application between layer1 (ethereum) and layer2 (zksync). I have successfully deployed 2 contracts FundCoreL1.sol and FundBridgeL1.sol on ethereum sepolia. However, when I deployed the remaining 2 contracts to zksync testnet, I encountered an error.

### Additional Details

kien@DESKTOP-PFDK8AA MINGW64 /d/Workspace/Project/zksync-project-specialize (master)
$ truffle migrate --network zkSyncTestnet --f 3 --to 3 --reset

Compiling your contracts...
===========================
> Compiling @matterlabs\zksync-contracts\l2\system-contracts\interfaces\IL1Messenger.sol
> Compiling @openzeppelin\contracts\access\Ownable.sol
> Compiling @openzeppelin\contracts\utils\Context.sol
> Compiling .\contracts\L2\FundBridgeL2.sol
> Compiling .\contracts\L2\FundCoreL2.sol
> Artifacts written to D:\Workspace\Project\zksync-project-specialize\build\contracts
> Compiled successfully using:
   - solc: 0.8.19+commit.7dd6d404.Emscripten.clang

Starting migrations...
======================
> Network name:    'zkSyncTestnet'
> Network id:      300
> Block gas limit: 1125899906842624 (0x4000000000000)

3_deploy_fund_core_l2.js
========================
Deploying FundCoreL2 with params: 0x75a768bBBE1e082d7Be9E390A6FCecF728B60d84 0x0000000000000000000000000000000000008008

   Deploying 'FundCoreL2'
   ----------------------
 *** Deployment Failed ***

"FundCoreL2" -- Failed to serialize transaction: toAddressIsNull.

Exiting: Review successful transactions manually by checking the transaction hashes above on Etherscan.

Error:  *** Deployment Failed ***

"FundCoreL2" -- Failed to serialize transaction: toAddressIsNull.

Error:  *** Deployment Failed ***

"FundCoreL2" -- Failed to serialize transaction: toAddressIsNull.

    at C:\Users\kien\AppData\Roaming\npm\node_modules\truffle\build\webpack:\packages\deployer\src\deployment.js:330:1
    at processTicksAndRejections (node:internal/process/task_queues:95:5)     
Truffle v5.11.5 (core: 5.11.5)
Node v20.16.0

**Comment by alhenaai** [comment]
The error message "Failed to serialize transaction: toAddressIsNull" typically indicates that the deployment script is trying to deploy a contract to an address that is not specified or is null. This can happen if the contract address or the constructor parameters for the contract are not correctly set in your deployment script.

Here are a few steps you can take to troubleshoot and resolve this issue:

1. **Check Constructor Parameters**: Ensure that all required parameters for the `FundCoreL2` contract are correctly passed in the deployment script. The error might be due to missing or incorrect constructor arguments.

2. **Verify Contract Address**: Make sure that the contract address you are deploying to is correctly specified. If this is a new deployment, ensure that the address is not being set from a previous deployment or an incorrect source.

3. **Review Deployment Script**: Look at your `3_deploy_fund_core_l2.js` script. Ensure that the script correctly specifies the address to which the contract should be deployed. If the address is derived from another contract or a configuration file, verify that these are correctly set up.

4. **Network Configuration**: Since you are deploying to the zkSync Testnet, ensure that your network configuration in Truffle is correctly set up to interact with zkSync's Layer 2. This includes setting the correct RPC URLs, chain IDs, and any other network-specific parameters.

5. **Truffle Configuration**: Check your `truffle-config.js` file to ensure that the network settings for `zkSyncTestnet` are correctly configured. This should include the correct provider and network ID.

6. **Contract Compilation**: Ensure that there are no issues during the compilation of your contracts that might affect deployment. Since the compilation seems successful, this is less likely the issue but always good to double-check.

If after these checks the issue persists, you might want to simplify your deployment script to deploy a minimal version of the contract to isolate the issue further. This can help determine if the problem is with the contract code or the deployment script/environment.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by bxpana** [reply, marked as answer]
@kientt265 Just a heads up, Truffle is planning on being discontinued soon so you'll want to look for another tool to use for deployment. 

We have some tutorials on deploying with Hardhat and Foundry in our docs that might help you with this issue too https://docs.zksync.io/build/start-coding/zksync-101

One thing that stands out is in your FundBridgeL2 contract you are importing Il1Messenger.sol from something that doesn't exist. There is no `zksync-contracts` repo in the Matter Labs org. Check out the examples in our tutorials and see if those help you out.
